File: scm-csharp/scheme/primitives/PrimitiveSysScmVersion.cs
namespace scheme;

using System.Reflection;

public class PrimitiveSysScmVersion : Primitive
{
    public override string Name() => "sys-scm-version";

    public override string Info() =>
        "Syntax: (sys-scm-version)\n" +
        "Library: (scm system)\n" +
        "Description: Returns the SCM interpreter version as a string.\n" +
        "Example:\n" +
        "  (sys-scm-version) => \"0.0.1\"";

    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 0, 0);
        var version = Assembly.GetExecutingAssembly()
            .GetName().Version?.ToString(3) ?? "unknown";
        return version.ToCharArray();
    }
}

File: scm-java/src/main/scheme/primitives/PrimitiveSysScmVersion.java
package scheme.primitives;

import scheme.*;
import java.io.InputStream;

public class PrimitiveSysScmVersion extends Primitive {
    @Override public String name() { return "sys-scm-version"; }

    @Override public String info() {
        return "Syntax: (sys-scm-version)\n" +
               "Library: (scm system)\n" +
               "Description: Returns the SCM interpreter version as a string.\n" +
               "Example:\n" +
               "  (sys-scm-version) => \"0.0.1\"";
    }

    @Override public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 0, 0);
        String version = "unknown";
        try (InputStream is = getClass().getResourceAsStream("/version.txt")) {
            if (is != null) {
                version = new String(is.readAllBytes()).trim();
            }
        } catch (Exception e) {}
        return version.toCharArray();
    }
}
